src/cljam/io/vcf/util.clj

```diff
@@ -286,3 +286,93 @@
+                         (map = (cstr/reverse upper-ref))
+                         (take-while true?)
+                         count
+                         long
+                         (Math/min (- (Math/min ref-length alt-length)
+                                      left-match)))
+        matched-length (+ left-match right-match)]
+    ;; Note that `{:type :ref}` is handled in the caller
+    (cond
+      (and (= left-match ref-length) (< left-match alt-length))
+      {:type :insertion, :offset (dec left-match),
+       :n-bases (- alt-length left-match), :inserted (subs alt left-match)}
+
+      (and (< left-match ref-length) (= left-match alt-length))
+      {:type :deletion, :offset (dec left-match),
+       :n-bases (- left-match ref-length), :deleted (subs ref left-match)}
+
+      (= (inc matched-length) ref-length alt-length)
+      {:type :snv, :ref (nth ref left-match),
+       :alt (nth alt left-match), :offset left-match}
+
+      (= matched-length alt-length)
+      {:type :deletion, :offset (dec left-match),
+       :n-bases (- left-match (- ref-length right-match)),
+       :deleted (subs ref left-match (- ref-length right-match))}
+
+      (= matched-length ref-length)
+      {:type :insertion, :offset (dec left-match),
+       :n-bases (- alt-length right-match left-match),
+       :inserted (subs alt left-match (- alt-length right-match))}
+
+      (= ref-length alt-length)
+      {:type :mnv, :offset left-match,
+       :ref (subs ref left-match (- ref-length right-match)),
+       :alt (subs alt left-match (- alt-length right-match))}
+
+      :else {:type :complex})))
+
+(defn inspect-allele
+  "Inspects an `alt` allele by comparing to a `ref` allele string.
+  Returns a map containing `:type` and other detailed information.
+  A value of the key `:type` can be one of the followings:
+  - `:ref` Non-variant allele
+  - `:id` Symbolic reference
+  - `:snv` Single nucleotide variant
+  - `:mnv` Multiple nucleotide variants
+  - `:insertion` Insertion of a short base sequence
+  - `:deletion` Deletion of a short base sequence
+  - `:complete-insertion` Complete insertion of a long sequence
+  - `:breakend` Breakend of a complex rearrangement
+  - `:complex` Complex nucleotide variants other than snv/mnv/indel
+  - `:other` Can't categorize the allele, might be malformed"
+  [ref alt]
+  (let [alt-length (count alt)]
+    (cond
+      (or (#{"." "*" "X" "<*>" "<X>"} alt)
+          (= (cstr/upper-case ref) (cstr/upper-case alt)))
+      {:type :ref}
+
+      (re-matches #"<.+>" alt)
+      {:type :id, :id (subs alt 1 (dec alt-length))}
+
+      (re-matches #"(?i)([ACGTN]<.+>|<.+>[ACGTN])" alt)
+      (if (cstr/starts-with? alt "<")
+        {:type :complete-insertion,
+         :join :before,
+         :base (last alt),
+         :id (subs alt 1 (- alt-length 2))}
+        {:type :complete-insertion,
+         :join :after,
+         :base (first alt),
+         :id (subs alt 2 (dec alt-length))})
```

Comment:
I think it's a better way to use capturing values in regexp (please correct me if I'm wrong).

```clojure
  [ref alt]
  (if (or (#{"." "*" "X" "<*>" "<X>"} alt)
          (= (cstr/upper-case ref) (cstr/upper-case alt)))
    {:type :ref}
    (condp re-matches alt
      #"<(.+)>" :>> (fn [[_ id]] {:type :id, :id id})
      #"(?i)<(.+)>([ACGTN])" :>> (fn [[_ id [base]]]
                                   {:type :complete-insertion,
                                    :join :before,
                                    :base base,
                                    :id id})
      #"(?i)([ACGTN])<(.+)>" :>> (fn [[_ [base] id]]
                                   {:type :complete-insertion,
                                    :join :after,
                                    :base base,
                                    :id id})
      (or (some-> (parse-breakend alt)
                  (assoc :type :breakend))
          (if (and (re-matches #"(?i)[ACGTN]+" ref)
                   (re-matches #"(?i)[ACGTN]+" alt))
            (inspect-nucleotides-allele ref alt)
            {:type :other})))))
```

Reply:
Thank you for suggesting `condp`!
 I rewrote the patch with it and also tweaked the categorization in dc63967.
I think it looks better now 😄